### PR TITLE
fix(velodyne_decoder): overflow handling in vls128

### PR DESCRIPTION
**Pull request by knzo25**
The timestamps in the vls128 were still relative to the last pointcloud, which was producing some errors Additionally, when entire scans were dropped, the overflow was not being handled correctly (this bug was also present in the awf velodyne driver)

## PR Type

- Bug Fix

## Related Links
[TIER IV INTERNAL LINK (discussion)](https://star4.slack.com/archives/C057JDFS5M2/p1703655034662449)
[TIER IV INTERNAL LINK (footage)](https://drive.google.com/drive/folders/1sB3iTCVo9Z4ysu-HwtBLWnPHRN3Vmk6_?usp=drive_link)

## Description
2 bugs were discovered and addressed related to the handling of time stamps during pointcloud overflow between scans.
Context: 
 - The lidar scan has an azimuth value (the origin is defined by the sensor) and usually a pointcloud would start at 0 degrees and end in 360 degrees. 
 - However, to synchronize multiple lidars and cameras we can actually make pointclouds  starting from a different angle.
 - Since the communication with the sensor is via "small" packets instead of a single packet per physical scan (360 degrees) we can decide how to assemble the packets into a pointcloud (e.g, start making a pointcloud starting from sensor angle = 30 degrees).
 - However, since each packet contains several points with different azimuth values, we need to handle one packet in two poinclouds (we call this concept overflow). Example: a certain packet contains points both lower and higher than a desired azimuth threshold. Lower values are added to the current pointcloud, whereas the higher ones need to be handled by the next one.

Nebula-only problem:
 - When a scan message arrives at the decoder, the absolute time stamp gets taken from the first block of the first package in the scan message. In nebula (velodyne_points), each individual point has a relative time stamp, and they are relative to the previously stated absolute time stamp.
 - However, the points that were determined to be part of the next pointcloud (overflow) currently retain their relative timestamp. Since they are the last points of the scan, they actually have relative times close to 100ms (the highest value, since physical scans run at 10hz).
 - When the next scan message arrives, the process is repeated once again, but this time, the first points of the pointcloud will have a very high relative time, followed by points with relative time close to 0 and increasing to almost 100ms (there is a discontinuity).

Although only the non-overflow points are OK, if we use algorithms that integrate timestamps from all the points starting from the beginning (overflow points), it will cause big errors. This was particularly observed in the distortion correction algorithm (see the provided videos)

The solution simply consists of using as global timestamp the stamp of the overflow points (some conversions were required).

Nebula & old vls driver problem:
 - In the vehicle, due to how pointcloud containers work, "there are no dropped packages". All packets from the lidar end up being converted into pointclouds.
 - However, when recording and replaying rosbags, some entire scans (aggregation of sensor packets corresponding to one physical 360 degree scan) can be missing.
 - In this case, even using the previews fix, errors still occur, since the overflow packets are used to determine the global time stamp, but since the points following the overflow are missing, the points after the overflow corresponds to the ones of the next physical scan, which makes a total time difference between the first and last point of the pointcloud of about 200ms (two scans).

For this problem, the proposed solution is that we first detect dropped packages using a rough stamp criteria, and then just ignore the last pointcloud's overflow when necessary (since there are abut 600 packages per scan, we are only loosing 360 * 1 / 600 degrees of information)

## Review Procedure
A good way to check that the timestamps are being computed correctly, is comparing the raw_pointcloud with the output of the distortion corrector of autoware.

For example, it would be convenient to compare the behavior of the output of the distortion correction before/after this PR
(additionally, the fix for the original vls implementation also serves as a good baseline https://github.com/knzo25/awf_velodyne/tree/fix/drop_packages_fix )

## Pre-Review Checklist for the PR Author

**PR Author should check the checkboxes below when creating the PR.**

- [x] Assign PR to reviewer

## Checklist for the PR Reviewer

**Reviewers should check the checkboxes below before approval.**

- [ ] Commits are properly organized and messages are according to the guideline
- [ ] (Optional) Unit tests have been written for new behavior
- [x] PR title describes the changes

## Post-Review Checklist for the PR Author

**PR Author should check the checkboxes below before merging.**

- [ ] All open points are addressed and tracked via issues or tickets

## CI Checks

- **Build and test for PR**: Required to pass before the merge.

### CODE REVIEW

File: nebula_decoders/src/nebula_decoders_velodyne/decoders/vls128_decoder.cpp [hunk 1]
```diff
@@ -75,24 +80,56 @@ std::tuple<drivers::NebulaPointCloudPtr, double> Vls128Decoder::get_pointcloud()
...
 {
+  if (overflow_pc_->points.size() == 0) {
+    scan_timestamp_ = -1;
+    overflow_pc_->points.reserve(max_pts_);
+    return;
+  }
+
+  // Compute the absolute time stamp of the last point of the overflow pointcloud
+  const double last_overflow_time_stamp =
+    scan_timestamp_ + 1e-9 * overflow_pc_->points.back().time_stamp;
```
[the remark this answers is not in the document]
Comment (bgilby59) [reply nebula_decoders/src/nebula_decoders_velodyne/decoders/vls128_decoder.cpp:107]: Or just add a comment that there is a ns to s conversion?

File: nebula_decoders/src/nebula_decoders_velodyne/decoders/vls128_decoder.cpp [hunk 2]
```diff
@@ -75,24 +80,56 @@ std::tuple<drivers::NebulaPointCloudPtr, double> Vls128Decoder::get_pointcloud()
...
   // Add the overflow buffer points
   for (size_t i = 0; i < overflow_pc_->points.size(); i++) {
-    scan_pc_->points.emplace_back(overflow_pc_->points[i]);
+    auto overflow_point = overflow_pc_->points[i];
+
+    // The overflow points had the stamps from the previous pointcloud. These need to be changed to
+    // be relative to the overflow's packet timestamp
+    double new_timestamp =
+      scan_timestamp_ + 1e-9 * overflow_point.time_stamp - last_block_timestamp_;
+    overflow_point.time_stamp = 1e-9 * new_timestamp;
```
[the remark this answers is not in the document]
Comment (knzo25) [reply nebula_decoders/src/nebula_decoders_velodyne/decoders/vls128_decoder.cpp]: Thanks for the catch !
Fixed in [cf0c795](https://github.com/tier4/nebula/pull/111/commits/cf0c79527f7c98e947b81c5100ceec9ed0ec7eb0)

File: nebula_decoders/src/nebula_decoders_velodyne/decoders/vls128_decoder.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (knzo25) [reply nebula_decoders/src/nebula_decoders_velodyne/decoders/vls128_decoder.cpp:107]: While I agree with improving readability, this conversion is all over the place in nebula, so for now I would prefer to leave it as is.

**Comment by knzo25** [comment]
@bgilby59 
Regarding the additional problems we experienced the other day, [9091b2f](https://github.com/tier4/nebula/pull/111/commits/9091b2fe40afbbb4fe8c812490eef4ab1faa102d) should fix those.

File: nebula_decoders/src/nebula_decoders_velodyne/decoders/vls128_decoder.cpp [hunk 3]
```diff
@@ -116,16 +116,17 @@ void Vls128Decoder::reset_overflow(double time_stamp)
   }
 
   // Add the overflow buffer points
-  for (size_t i = 0; i < overflow_pc_->points.size(); i++) {
-    auto overflow_point = overflow_pc_->points[i];
+  while (overflow_pc_->points.size() > 0) {
+    auto overflow_point = overflow_pc_->points.back();
```
Comment (bgilby59) [inline nebula_decoders/src/nebula_decoders_velodyne/decoders/vls128_decoder.cpp:120]: Why did you change the for loop to a while loop? Looks fine, but I'm just curious about the motivation to make the change.

Comment (knzo25) [reply nebula_decoders/src/nebula_decoders_velodyne/decoders/vls128_decoder.cpp:120]: Sorry, forgot to mention it before.
If you look at the way the overflow_pc if constructed, the order of the points is reversed.
In the implementation prior to [9091b2f](https://github.com/tier4/nebula/pull/111/commits/9091b2fe40afbbb4fe8c812490eef4ab1faa102d) the new order was kept and copied into the new pc. That is, the output pc was reversed.

While checking the timestamps of the pc I realized that the timestamps were not in ascending order, which led me to find the error. That is way, by reversing the order again while moving the overflow into the new pc, the "correct" order is preserved

Comment (bgilby59) [reply nebula_decoders/src/nebula_decoders_velodyne/decoders/vls128_decoder.cpp:120]: Ah, I see now. The previous implementation was using `emplace_back()`. Thanks!

File: nebula_decoders/src/nebula_decoders_velodyne/decoders/vls128_decoder.cpp [hunk 4]
```diff
@@ -75,24 +80,57 @@ std::tuple<drivers::NebulaPointCloudPtr, double> Vls128Decoder::get_pointcloud()
...
-  for (size_t i = 0; i < overflow_pc_->points.size(); i++) {
-    scan_pc_->points.emplace_back(overflow_pc_->points[i]);
+  while (overflow_pc_->points.size() > 0) {
+    auto overflow_point = overflow_pc_->points.back();
+
+    // The overflow points had the stamps from the previous pointcloud. These need to be changed to
+    // be relative to the overflow's packet timestamp
+    double new_timestamp_seconds =
+      scan_timestamp_ + 1e-9 * overflow_point.time_stamp - last_block_timestamp_;
+    overflow_point.time_stamp = new_timestamp_seconds < 0.0 ? 0.0 : 1e9 * new_timestamp_seconds;
```
Comment (bgilby59) [inline nebula_decoders/src/nebula_decoders_velodyne/decoders/vls128_decoder.cpp:126]:
```suggestion
    overflow_point.time_stamp = static_cast<uint32_t>(new_timestamp_seconds < 0.0 ? 0.0 : 1e9 * new_timestamp_seconds);
```

File: nebula_decoders/src/nebula_decoders_velodyne/decoders/vls128_decoder.cpp [hunk 5]
```diff
@@ -75,24 +80,57 @@ std::tuple<drivers::NebulaPointCloudPtr, double> Vls128Decoder::get_pointcloud()
...
+    scan_timestamp_ = -1;
+    overflow_pc_->points.reserve(max_pts_);
+    return;
+  }
+
+  // Compute the absolute time stamp of the last point of the overflow pointcloud
+  const double last_overflow_time_stamp =
+    scan_timestamp_ + 1e-9 * overflow_pc_->points.back().time_stamp;
+
+  // Detect cases where there in an unacceptable time difference between the last overflow point and
```
Comment (bgilby59) [inline nebula_decoders/src/nebula_decoders_velodyne/decoders/vls128_decoder.cpp:109]: Typo fix:
```suggestion
  // Detect cases where there is an unacceptable time difference between the last overflow point and
```

**Review by bgilby59 — APPROVED** [review]
Looks good. Tested on my end with 3 rosbags with XX1 data. The rosbags used and videos of results can be found at the below link:

[TIER IV INTERNAL LINK](https://drive.google.com/drive/folders/1DtnhiGz5xRRozB9l0LVhHEL9ceG5xh_4)
